### PR TITLE
Update `m2ctx.py`

### PR DESCRIPTION
**Pull request by AngheloAlf**
I cleaned up the script and modified it so macros are preserved. Useful for cases when you want to manually create a decomp.me scratch because gfx macros are preserved in context

### CODE REVIEW

**Review by hensldm — APPROVED** [review]
Works fine for me :)

File: tools/m2ctx.py [hunk 1]
```diff
@@ -1,14 +1,43 @@
...
-src_dir = root_dir + "src/"
+script_dir = Path(os.path.dirname(os.path.realpath(__file__)))
+root_dir = script_dir / ".."
+src_dir = root_dir / "src"
+
+# Project-specific
+CPP_FLAGS = [
+    "-Iinclude",
+    "-Iassets",
+    "-Isrc",
```
Comment (hensldm) [inline tools/m2ctx.py:18]: can you add `Ibuild` and `I.`, as well switch `Iassets` and `Isrc` to fully match with the Makefile?
`-Iinclude -Isrc -Iassets -Ibuild -I.`

**Review by hensldm — CHANGES_REQUESTED** [review]

**Review by hensldm — APPROVED** [review]

**Review by isghj5 — APPROVED** [review]

File: tools/m2ctx.py [hunk 2]
```diff
@@ -1,14 +1,45 @@
...
+    "-D_LANGUAGE_C",
+    "-DNON_MATCHING",
+    "-D_Static_assert(x, y)=",
+    "-D__attribute__(x)="
+    "-DNDEBUG",
+    "-D_FINALROM",
+    "-D_MIPS_SZLONG=32",
+    "-DSCRIPT(x)="
+    "-D__attribute__(x)=",
+    "-D__asm__(x)=",
```
Comment (EllipticEllipsis) [inline tools/m2ctx.py:32]: This one is completely useless since we'll never use __asm__ when matching.

File: tools/m2ctx.py [hunk 3]
```diff
@@ -1,14 +1,45 @@
...
+    "-Ibuild",
+    "-I.",
+
+    "-D__sgi",
+    "-D_LANGUAGE_C",
+    "-DNON_MATCHING",
+    "-D_Static_assert(x, y)=",
+    "-D__attribute__(x)="
+    "-DNDEBUG",
+    "-D_FINALROM",
```
Comment (EllipticEllipsis) [inline tools/m2ctx.py:28]: Maybe, maybe not.

File: tools/m2ctx.py [hunk 4]
```diff
@@ -1,14 +1,45 @@
...
+    "-D_MIPS_SZLONG=32",
+    "-DSCRIPT(x)="
```
Comment (EllipticEllipsis) [inline tools/m2ctx.py:30]: ?

File: tools/m2ctx.py [hunk 5]
```diff
@@ -48,63 +78,55 @@ def custom_replacements(output):
...
+def import_c_file(in_file) -> str:
+    in_file = os.path.relpath(in_file, root_dir)
 
-def get_c_file(directory):
-    for root, dirs, files in os.walk(directory):
-        for file in files:
-            if file.endswith(".c") and "data" not in file:
-                return file
+    cpp_command = ["gcc", "-E", "-P", "-undef", "-dM", *CPP_FLAGS, in_file]
+    cpp_command2 = ["gcc", "-E", "-P", "-undef", *CPP_FLAGS, in_file]
```
Comment (EllipticEllipsis) [inline tools/m2ctx.py:85]: Why are there two of these?

Comment (AngheloAlf) [reply tools/m2ctx.py:85]: The first one has `-dM` which is the thing that does the macro preservation

**Review by EllipticEllipsis — APPROVED** [review]

**Pull request merged**